src/sparse/KokkosSparse_BlockCrsMatrix.hpp

```diff
@@ -0,0 +1,940 @@
+  }
+
+  // Return unmanaged 2D strided View wrapping local block K from this block-row
+  KOKKOS_INLINE_FUNCTION
+  block_values_type block(const ordinal_type& K) const {
+    return block_values_type( &(values_[K*blockDim_]), Kokkos::LayoutStride(blockDim_,length*blockDim_,blockDim_,1) );
+  }
+
+
+  KOKKOS_INLINE_FUNCTION
+  ordinal_type findRelBlockOffset ( const ordinal_type &idx_to_match, bool is_sorted = false ) const {
+    typename std::remove_cv<ordinal_type>::type offset = std::numeric_limits< ordinal_type >::max();
+    for ( typename std::remove_cv<ordinal_type>::type blk_offset = 0; blk_offset < length; ++blk_offset ) {
+      ordinal_type idx = colidx_[blk_offset];
+      if ( idx == idx_to_match ) { offset = blk_offset; } // return relative offset
+      //if ( idx == idx_to_match ) { offset = blk_offset*blockDim_; } // return offset into values_
+    }
+    return offset;
+  }
+
+};
+
+/// \class BlockCrsMatrix
+/// \brief Compressed sparse row implementation of a sparse matrix.
+/// \tparam ScalarType The type of entries in the sparse matrix.
+/// \tparam OrdinalType The type of column indices in the sparse matrix.
+/// \tparam Device The Kokkos Device type.
+/// \tparam MemoryTraits Traits describing how Kokkos manages and
+///   accesses data.  The default parameter suffices for most users.
+///
+/// "Crs" stands for "compressed row sparse."  This is the phrase
+/// Trilinos traditionally uses to describe compressed sparse row
+/// storage for sparse matrices, as described, for example, in Saad
+/// (2nd ed.).
+template<class ScalarType,
+         class OrdinalType,
+         class Device,
+         class MemoryTraits = void,
+         class SizeType = typename Kokkos::ViewTraits<OrdinalType*, Device, void, void>::size_type>
+class BlockCrsMatrix {
+private:
+  typedef typename Kokkos::ViewTraits<ScalarType*,Device,void,void>::host_mirror_space host_mirror_space ;
+public:
+  //! Type of the matrix's execution space.
+  typedef typename Device::execution_space execution_space;
+  //! Type of the matrix's memory space.
+  typedef typename Device::memory_space memory_space;
+  //! Type of the matrix's device type.
+  typedef Kokkos::Device<execution_space, memory_space> device_type;
+
+  //! Type of each value in the matrix.
+  typedef ScalarType value_type;
+  //! Type of each (column) index in the matrix.
+  typedef OrdinalType ordinal_type;
+  typedef MemoryTraits memory_traits;
+  /// \brief Type of each entry of the "row map."
+  ///
+  /// The "row map" corresponds to the \c ptr array of row offsets in
+  /// compressed sparse row (CSR) storage.
+  typedef SizeType size_type;
+
+  //! Type of a host-memory mirror of the sparse matrix.
+  typedef BlockCrsMatrix<ScalarType, OrdinalType, host_mirror_space, MemoryTraits> HostMirror;
+  //! Type of the graph structure of the sparse matrix.
+  typedef Kokkos::StaticCrsGraph<OrdinalType, Kokkos::LayoutLeft, execution_space, SizeType> StaticCrsGraphType;
```

Comment:
We should take this opportunity to make this typedef have consistent formatting with the other typedefs.  You could always keep this inconsistently formatted typedef around as an extra backwards compatibility measure with `KokkosSparse::CrsMatrix`, but that one really should change too.

Reply:
I added an extra typedef consistent with the one in Kokkos: `staticcrsgraph_type`, but left this one for backward compatibility like you mentioned in case someone is using it outside of Trilinos.